accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+
+**Conclusion**
+
+No build magic and more shared code prospect, as well as the ability to test HttpClient changes easily across all platforms, are very appealing.
+
+## The plan for .NET 5
+
+The decision is to use different approaches for each platform for .NET 5. We might consider unifying the approaches eventually but not in .NET 5 time-frame.
+
+A new feature like setting will be introduced which will allow controlling the implicit handler selection. The internal property will be available inside System.Net.Http namespace and mapped to msbuild property or AppContext setting.
+
+```c#
+namespace System.Net.Http
+​​{
+​​  public partial class HttpClientHandler
+​​  {
+​​    internal static HandlerType DefaultHandlerType { get; } = HttpClientHandlerType.PlatformNative;
+​​  }
+}
+```
+
+The implementation has to use the property in the way to allow the ILLinker to remove the unnecessary dependency of unused handler when the type is not used elsewhere.
+
+### Xamarin.iOS/Xamarin.tvOS
+Follow proposal #2b and hook up existing `NSUrlSessionHandler` using `Activator.CreateInstance ("Foundation.NSUrlSessionHandler, Xamarin.iOS/Xamarin.tvOS")​`. The actual implementation will remain inside xamarin/xamarin-macios​ in repo.
+
+The plan ignores CFNetworkHandler and other legacy handlers. They will stay in `xamarin/xamarin-macios` and will be available only when constructed manually. The implementation logic inside HttpClientHandler will rely on DefaultHandlerType selection to set the implicit handler behaviour between NSUrlSessionHandler and SocketHttpHandler.
+
+### Xamarin.Mac
+There won’t be any platform specific handlers implicitly called by HttpClient OSX RID build of libraries (Runtime Pack). The existing SocketHttpHandler is what .NET Core supports today and have all features available for cross platform support (e.g. gRPC) for .NET. Xamarin.Mac could still ship any platform specific handler but they won’t be used unless explicitly constructed and passed to `HttpClient​` constructor.
+
+### Xamarin.Android
+
+Follow proposal #2b and hook up existing AndroidClientHandler using `Activator.CreateInstance (``"``Xamarin.Android.Net.AndroidClientHandler, Mono.Android``"``)`. The whole AndroidClientHandler implementation will remain inside `xamarin/xamarin-android`.
+
+Only RID specific version of System.Net.Http will be built as there won’t be any new API available as part of `dotnet/runtime` version.
+
+The code will most likely have to include special linker annotations to keep the dependency on the constructor inside external “unknown” assembly.
+
+### WebAssembly
+Follow proposal #5 and port WasmHttpMessageHandler to dotnet/runtime together with wasm interop layer.
+
+The public implementation of WasmHttpMessageHandler will be packaged into special NuGet which will be created to hold only this public type. The interop layer needed to support this handler will be also included into the NuGet package. To match existing naming it will be called
```

Comment:
Based on the API review we did of WasmHttpMessageHandler, this case may end up being revisited, mainly because the settings it was exposing are meant to change per request.

Reply:
The design here does not go into such details so that should be fine but I updated the names to reflect the latest naming